tsl/src/remote/copy_fetcher.c

```diff
@@ -286,32 +344,21 @@ copy_fetcher_complete(CopyFetcher *fetcher)
 
 			StringInfoData copy_data = { 0 };
 
-			copy_data.len = PQgetCopyData(remote_connection_get_pg_conn(fetcher->state.conn),
+			copy_data.len = PQgetCopyData(conn,
 										  &copy_data.data,
 										  /* async = */ false);
 
+			/* Set dataptr to ensure data is freed with PQfreemem() in
+			 * PG_CATCH() clause in case error is thrown. */
+			dataptr = copy_data.data;
+
 			if (copy_data.len == -1)
 			{
-				/*
-				 * According to the docs, this means EOF, but in practice we get
-				 * it when there is an error. Check for error and report it.
-				 */
+				/* Note: it is possible to get EOF without having received the
+				 * file trailer in case there's e.g., a remote error. */
 				fetcher->state.eof = true;
-
-				PGresult *res = PQgetResult(remote_connection_get_pg_conn(fetcher->state.conn));
-				if (res == NULL)
-				{
-					/* Shouldn't really happen but technically possible. */
-					TSConnectionError err;
-					remote_connection_get_error(fetcher->state.conn, &err);
-					remote_connection_error_elog(&err, ERROR);
-				}
-				if (PQresultStatus(res) != PGRES_COPY_OUT)
-				{
-					TSConnectionError err;
-					remote_connection_get_result_error(res, &err);
-					remote_connection_error_elog(&err, ERROR);
-				}
+				/* Should read final result with PQgetResult() until it
+				 * returns NULL. This happens when the fetcher is closed. */
```

Comment:
Looks like it will happen in line 497 instead (the call to `end_copy()` at the end of the top-level loop).